openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -768,6 +768,87 @@ def _get_encrypted_request_headers(request):
+    are invalidated when a user changes their email in one browser.
+    The active session in which the email change is made will remain valid.
+
+    The user's email is stored in their session during login
+    and gets updated when the user changes their email.
+    This middleware checks for any mismatch between the stored email
+    and the current user's email in each request, and if found,
+    it invalidates/flushes the session and mark cookies for deletion in request
+    which are then deleted in the process_response of SafeSessionMiddleware.
+    """
+
+    def process_request(self, request):
+        """
+        Invalidate the user session if there's a mismatch
+        between the email in the user's session and request.user.email.
+        """
+        if request.user.is_authenticated:
+            user_session_email = request.session.get('email', None)
+            are_emails_mismatched = user_session_email is not None and request.user.email != user_session_email
+            EmailChangeSessionInvalidationMiddleware._set_session_email_match_custom_attribute(are_emails_mismatched)
+            if settings.ENFORCE_SESSION_EMAIL_MATCH and are_emails_mismatched:
+                # Flush the session and mark cookies for deletion.
+                request.session.flush()
+                request.user = AnonymousUser()
+                _mark_cookie_for_deletion(request)
+
+    def process_response(self, request, response):
+        """
+        Update the logged-in cookies if the email change was marked
+        """
+        if (
+            settings.ENFORCE_SESSION_EMAIL_MATCH
+            and request.user.is_authenticated
+            and self._is_email_change_marked(request)
+        ):
+            # Update the JWT cookies with new user email
+            response = set_logged_in_cookies(request, response, request.user)
+
+        return response
+
+    @staticmethod
+    def register_email_change(request, email):
+        """
+        Sets email in session for comparison and mark session and user email change for cookie updates
+        """
+        request.session['email'] = email
+        EmailChangeSessionInvalidationMiddleware._mark_email_change(request)
+
+    @staticmethod
+    def _set_session_email_match_custom_attribute(value):
+        """
+        Sets custom attribute of session_email_match
+        """
+        # .. custom_attribute_name: session_email_match
+        # .. custom_attribute_description: Indicates whether there is a match between the
+        #      email in the user's session and the current user's email in the request.
+        set_custom_attribute('session_email_match', value)
+
+    @staticmethod
+    def _mark_email_change(request):
+        """
+        Sets a variable in the given request object to designate that the email
+        is changed. It will be used in process_response for updating the cookies
+        """
+        request.email_changed = True
```

Comment:
* [nit] This is just an opinion, but I'd drop this method and just move this to `register_email_change`. It just made me hunt to see that that was the only caller.
* [nit] We usually use the RequestCache for data like this. Note that if we use the RequestCache, it needs to be cleared before each test. You can see examples in other test files that use this.